feat: support blpop/brpop

cheniujh opened the pull request:
新增对redis的blpop、brpop命令的支持。 

**功能说明**：blpop/brpop是lpop/rpop的阻塞版本
- **阻塞功能**：如果用户从目标list中pop不到元素，就陷入阻塞，直到得到元素为止
- **支持超时机制**：用户可以指定最大阻塞等待时间（timeout）, 如果到达指定时间还未得到元素，返回nil
- **多list支持**：一个client可以同时阻塞等待多个目标list key, 一个list key 也能同时被多个client阻塞等待 

(redis官方文档示例：)
![image](https://github.com/OpenAtomFoundation/pika/assets/41671101/d1ef6dde-d649-489d-8af9-618627d48b55)

<br> 

**实现思路**：
1.维护了一个map，用于记录key与对应的阻塞客户端，大致如下:
<img width="941" alt="image" src="https://github.com/OpenAtomFoundation/pika/assets/41671101/24b56134-bdde-4719-a180-0ce07e232491">


2. 另外加了一个辅助的map，可以避免一些不必要的遍历：
<img width="1035" alt="image" src="https://github.com/OpenAtomFoundation/pika/assets/41671101/859a8fad-3924-4d2f-bc7c-9b05783965d1">

这两个map以及一个对应的std::shared_mutex加到了DispatchThread中作为成员变量。

**对这两个map的操作：**
1. **阻塞客户端**：当blpop/brpop无法从目标list中取得元素时，往maps_from_keys_to_conns中对应key的list中新增Node（也往map_from_conns_to_keys插入一个list of keys)。

2. **解阻塞**：当lpush,rpush，rpoplpush命令执行后，用maps_from_keys_to_conns执行一下find方法，看是否有客户端正在等待当前key（刚刚发生push行为的），如果有，则给往线程池塞一个任务，用于向被阻塞的客户端写响应，以及清理两个map中的对应项。

3. **超时扫描**：在serverThread的threadmain的epoll里注册定时任务（目前是200ms执行一次），每次扫描整个maps_from_keys_to_conns，查看是否有过期的Node，如果有，回写一个-1(nil)，并且清理两个map中的相关项。

4.  **断联清理**：当客户端断开连接时，对maps_from_conns_to_keys执行一下find，查看一下该连接是否是从被blpop/brpop阻塞的状态中退出的，如果是，则清理map。

5. **连接保活**：在检查客户端keep_alive的部分（默认是60s超时就会断开），加了一个判断：如果在maps_from_conns_to_keys中能find到该客户端的对应记录，则不断开此连接，且更新其last_interaction属性





**命令的行为：**
<img width="544" alt="image" src="https://github.com/OpenAtomFoundation/pika/assets/41671101/a4abb469-86c0-4d06-8060-06fb4fdb2bef">



**TODO:**

- [x]  blpop这种阻塞式命令对binlog有影响，需要调整
      **何时写binlog**：在blpop，brpop中重写了doBinlog方法，只有真正发生写行为的时候才写一条binlog
      **binlog写什么**：对于blpop/brpop来说，当用户真正得到元素时，往binlog写一条对应的lpop/rpop，如果用户在得到元素前就主动断开，则不写binlog（因为pop行为实际上也没有发生）。
    **产生的binlog**：命令部分由小写字母组成（lpop/rpop), 而用户执行Lpop/Rpop命令产生的binlog中的命令都是大写（LPOP/RPOP），大小写不影响从节点消费，同时也可以作为一种区别方式（区分该条lpop/rpop binlog究竟是blpop/brpop产生的，还是用户执行Lpop/Rpop命令产生的）


- [x] 单测：已集成到Github Workflow中，每次pr都会触发。
  -  覆盖了redis单测中对blpop/brpop的用例，并且额外增加了不少用例， 测试时分别开了address/thread sanitizer，全部通过
       - [x]   单个list不阻塞时的出列顺序（行为应当和lpop/rpop一样）
       - [x]   解阻塞测试（超时自动解阻塞，lpush解阻塞，rpush解阻塞，rpoplpush解阻塞）
       - [x]   并发下的阻塞/解阻塞测试
       - [x]   blpop/brpop多个list不阻塞时,从左到右选择第一个有元素的list进行pop
       - [x]   blpop/brpop的key中有重复key以及连续对同样的key进行blpop/brpop
       - [x]   阻塞客户端的目标list被一条push增加了多个value，先完成多个value的入列再pop
       - [x]   先被阻塞的先服务/阻塞最久的优先级最高
       - [x]  主从复制测试（备注：rpoplpush命令本身主从同步有问题，所以测试里把对应部分先注释了）
       - [x]  上述所有测试的多db并发运行
 
- [x]  代码整理：
              -  把方法体从.h移动到对应的.cpp
              -  删除测试用的打印输出
              -  使用clang-format整理格式(只整理了这个pr涉及到的文件）
 

> 给不应该被继承的cmd类都加上final关键字:  这个涉及到的类比较多，为了不对这个pr的可读性产生太大影响，决定下次提小pr的时候再一起做了（或者起个issue，也适合作为社区新人的first issue）

Fixes: #114

cheniujh committed:
working on BlpopCmd::Doinitial

cheniujh committed:
working on : making the maps and new monitoring thread

cheniujh committed:
added bLRPop_blocking_info_ in pika_server.cc

cheniujh committed:
adding functions

cheniujh committed:
adding functions02

cheniujh committed:
temp save

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into add_blpop_brpop

cheniujh committed:
data structure done

cheniujh committed:
working

cheniujh committed:
revised data structure

cheniujh committed:
working on moving

cheniujh committed:
working

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into add_blpop_brpop

cheniujh committed:
next step is to add a timeout scan

cheniujh committed:
added TimerTaskManager

cheniujh committed:
modified TimerTaskManager(v2)

cheniujh committed:
scan added

cheniujh committed:
tiny fix

cheniujh committed:
changed the task of unblocking conn from sync to async

cheniujh committed:
Merge branch 'back_up_add_blpop_brpop' of github.com:cheniujh/pika into new_unstable

# Conflicts:
#	src/net/src/dispatch_thread.cc
#	src/pika_command.cc
#	src/pstd/include/env.h

cheniujh committed:
removed some code for testing

wanghenshui commented on include/pika_list.h:
```diff
@@ -73,6 +74,29 @@ class LLenCmd : public Cmd {
   virtual void DoInitial() override;
 };
 
+class BLRPopBaseCmd : public Cmd {
```
叫BlockingBaseCmd比较合适，其他block命令也可以用，比如bzpopmin bzpopmax brpoplpush

cheniujh replied:
好的

wanghenshui commented on include/pika_list.h:
```diff
@@ -73,6 +74,29 @@ class LLenCmd : public Cmd {
+  int64_t expire_time_{0};
+};
+
+class BLPopCmd : public BLRPopBaseCmd {
```
class BLPopCmd final
加上final，一块儿全改了吧

cheniujh replied:
好的

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -55,6 +95,84 @@ class DispatchThread : public ServerThread {
+   */
+  void CleanWaitNodeOfUnBlockedBlrConn(std::shared_ptr<net::RedisConn> conn_unblocked) {
+    // removed all the waiting info of this conn/ doing cleaning work
+    auto& blpop_keys_list = map_from_conns_to_keys_for_blrpop.find(conn_unblocked->fd())->second;
```
如果find为空你这么写会core。要检查iter

cheniujh replied:
好的，我加个判断 （这个CleanWaitNodeOfUnBlockedBlrConn只在这个fd一定存在对应的pair时才调用，如果这里find没命中说明其他地方出错了，不过确实没有显式地说明过这点。

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -55,6 +95,84 @@ class DispatchThread : public ServerThread {
+    // removed all the waiting info of this conn/ doing cleaning work
+    auto& blpop_keys_list = map_from_conns_to_keys_for_blrpop.find(conn_unblocked->fd())->second;
+    for (auto& blpop_key : *blpop_keys_list) {
+      auto& wait_list_of_this_key = map_from_keys_to_conns_for_blrpop.find(blpop_key)->second;
```
同上，如果find为空你这么写会core。要检查iter

cheniujh replied:
好的

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -55,6 +95,84 @@ class DispatchThread : public ServerThread {
+      auto& wait_list_of_this_key = map_from_keys_to_conns_for_blrpop.find(blpop_key)->second;
+      for (auto conn = wait_list_of_this_key->begin(); conn != wait_list_of_this_key->end();) {
+        if (conn->GetConnBlocked()->fd() == conn_unblocked->fd()) {
+          wait_list_of_this_key->erase(conn);
```
你这种写法是map经典错误了，iter遍历中修改

得改成
conn = wait_list_of_this_key->erase(conn);

cheniujh replied:
本来我写的是conn = wait_list_of_this_key->erase(conn);，但是下一句就是break，所以犹豫一下又去掉了， 那我再加回去

wanghenshui commented:
还需要补充driver测试，用redis-py redisgo都行

cheniujh commented:
> 还需要补充driver测试，用redis-py redisgo都行

好的，不过这几天有些事，可能要过个几天再加上

AlexStocks commented:
@cheniujh please fix the file confliction.

cheniujh committed:
pull from unstable and solve conflicts

cheniujh committed:
change from partition to db

cheniujh committed:
temp save for testing

cheniujh committed:
handle binlog of blr/pop(when conn get served, write a binlog of lpop/rpop)

cheniujh commented on CMakeLists.txt:
不小心传了cmakelists，后面会删掉这个变动，包括一些测试代码也是，最后会删除掉

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into new_add_blpop_brpop

cheniujh committed:
improved sanitizer options in CMakeLists

cheniujh committed:
temp save

cheniujh committed:
add unit test

cheniujh committed:
modified cmakelists

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +82,114 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+  std::shared_ptr<Slot> slot;
+  net::DispatchThread* dispatchThread;
+  BlrPopUnblockTaskArgs(std::string  key_, std::shared_ptr<Slot> slot_, net::DispatchThread* dispatchThread_)
+      : key(std::move(key_)), slot(std::move(slot_)), dispatchThread(dispatchThread_){}
```
不要move shared_ptr

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +82,114 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+  auto dispatchThread = dynamic_cast<net::DispatchThread*>(curr_conn->thread());
+
+  {
+    std::shared_lock read_latch(dispatchThread->GetBLRPopBlockingMapLatch());
```
你这个锁太大了。尽量分片

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +82,114 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+    auto& map_from_keys_to_conns_for_blrpop = dispatchThread->GetMapFromKeysToConnsForBlrpop();
+    net::BlrPopKey blrPop_key{curr_conn->GetCurrentTable(), key};
+    auto it = map_from_keys_to_conns_for_blrpop.find(blrPop_key);
+    if (it == map_from_keys_to_conns_for_blrpop.end()) {
```
```c++
if ( auto it = map_from_keys_to_conns_for_blrpop.find(blrPop_key), it == map_from_keys_to_conns_for_blrpop.end()) 
```

可以写成一行，另外map_from_keys_to_conns_for_blrpop变量名字用驼峰式，且这个名字也太长了

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +82,114 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+void BlockingBaseCmd::ServeAndUnblockConns(void* args) {
+  auto bg_args = std::unique_ptr<BlrPopUnblockTaskArgs>(static_cast<BlrPopUnblockTaskArgs*>(args));
+  net::DispatchThread* dispatchThread = bg_args->dispatchThread;
+  std::shared_ptr<Slot> slot = std::move(bg_args->slot);
```
不要move shared_ptr

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +82,114 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+  std::string value;
+  rocksdb::Status s;
+  // traverse this list from head to tail(in the order of adding sequence) ,means "first blocked, first get served“
+  for (auto conn_blocked = waitting_list_of_this_key->begin(); conn_blocked != waitting_list_of_this_key->end();) {
```
用range for

cheniujh replied:
这里面有一句conn_blocked = waitting_list_of_this_key->erase(conn_blocked); ，所以没写rangefor

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +82,114 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+    std::string pop_type;
+    if(pop_arg.block_type == net::BlockPopType::Blpop){
+      pop_type = kCmdNameLPop;
+    }else if(pop_arg.block_type == net::BlockPopType::Brpop){
```
net::BlockPopType::Brpop 这玩意别放net里

wanghenshui commented on src/pika_list.cc:
```diff
@@ -95,6 +209,121 @@ void LPushCmd::Do(std::shared_ptr<Slot> slot) {
+
+  auto dispatchThread = dynamic_cast<net::DispatchThread*>(conn_to_block->thread());
+  std::lock_guard latch(dispatchThread->GetBLRPopBlockingMapLatch());
+  auto& map_from_keys_to_conns_for_blrpop = dispatchThread->GetMapFromKeysToConnsForBlrpop();
```
你这个变量命名我眼睛都花了。简单点求求了

wanghenshui commented on src/pika_list.cc:
```diff
@@ -95,6 +209,121 @@ void LPushCmd::Do(std::shared_ptr<Slot> slot) {
+      conn_to_block->fd(), std::make_unique<std::list<net::BlrPopKey>>(blrpop_keys.begin(), blrpop_keys.end()));
+
+  //------------code for testing---------------
+  std::cout << "-------------db name:" << conn_to_block->GetCurrentTable() << "-------------" << std::endl;
```
LOG(DEBUG)

wanghenshui commented on src/pika_list.cc:
```diff
@@ -95,6 +209,121 @@ void LPushCmd::Do(std::shared_ptr<Slot> slot) {
+  std::cout << "-----------end------------------" << std::endl;
+  //------------code for testing---------------
+}
+void BlockingBaseCmd::removeDuplicates(std::vector<std::string>& keys_) {
```
你这玩意用remove_if不行么

wanghenshui commented on src/pika_list.cc:
```diff
@@ -95,6 +209,121 @@ void LPushCmd::Do(std::shared_ptr<Slot> slot) {
+    res_.SetRes(CmdRes::kInvalidInt);
+    return;
+  }
+  constexpr int64_t seconds_of_ten_years = 10 * 365 * 24 * 3600;
```
这玩意单独抽出来放在开头

wanghenshui commented on src/pstd/include/env.h:
```diff
@@ -156,5 +159,104 @@ class RandomRWFile : public pstd::noncopyable {
+  std::function<void()> fun;
+} TimedTask;
+
+class TimedTaskManager {
```
这玩意放到net里

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -22,6 +26,51 @@ class NetItem;
 class WorkerThread;
 
+enum BlockPopType { Blpop, Brpop };
+typedef struct blrpopKey {  // this data struct is made for the scenario of multi dbs in pika.
```
可以叫BlockKey,不用写typedef

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -22,6 +26,51 @@ class NetItem;
 class NetFiredEvent;
 class WorkerThread;
 
+enum BlockPopType { Blpop, Brpop };
```
BlockKeyType

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -22,6 +26,51 @@ class NetItem;
+  std::string key;
+  bool operator==(const blrpopKey& p) const { return p.db_name == db_name && p.key == key; }
+} BlrPopKey;
+struct BlrPopKeyHash {
```
BlockKeyHash

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -22,6 +26,51 @@ class NetItem;
+  }
+};
+
+class BlockedPopConnNode {
```
BlockConnNode

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -22,6 +26,51 @@ class NetItem;
+class BlockedPopConnNode {
+ public:
+  virtual ~BlockedPopConnNode() {
+    std::cout << "~BlockedPopConnNode(): fd-" << conn_blocked_->fd() << " expire_time_:" << expire_time_ << std::endl;
```
LOG(DEBUG)

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -55,6 +104,89 @@ class DispatchThread : public ServerThread {
+    return map_from_keys_to_conns_for_blrpop_;
+  }
+  std::unordered_map<int, std::unique_ptr<std::list<BlrPopKey>>>& GetMapFromConnsToKeysForBlrpop() {
+    return map_from_conns_to_keys_for_blrpop_;
```
connToKey

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -55,6 +104,89 @@ class DispatchThread : public ServerThread {
+
+  std::unordered_map<BlrPopKey, std::unique_ptr<std::list<BlockedPopConnNode>>, BlrPopKeyHash>&
+  GetMapFromKeysToConnsForBlrpop() {
+    return map_from_keys_to_conns_for_blrpop_;
```
KeyToConn

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -55,6 +104,89 @@ class DispatchThread : public ServerThread {
+  }
+  std::shared_mutex& GetBLRPopBlockingMapLatch() { return bLRPop_blocking_map_latch_; };
+
+  pstd::TimedTaskManager& GetTimedTaskManager() { return timedTaskManager_; }
```
返回指针

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -71,6 +203,31 @@ class DispatchThread : public ServerThread {
+   *  were blocked by command blpop/brpop with key (eg. "list1").
+   */
+  std::unordered_map<BlrPopKey, std::unique_ptr<std::list<BlockedPopConnNode>>, BlrPopKeyHash>
+      map_from_keys_to_conns_for_blrpop_;
```
keyToConn

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -71,6 +203,31 @@ class DispatchThread : public ServerThread {
+   *  map_from_conns_to_keys_for_blrpop_:
+   *  mapping from conn(fd) to a list of keys that the client is waiting for.
+   */
+  std::unordered_map<int, std::unique_ptr<std::list<BlrPopKey>>> map_from_conns_to_keys_for_blrpop_;
```
connToKey

wanghenshui commented on src/net/src/dispatch_thread.h:
```diff
@@ -71,6 +203,31 @@ class DispatchThread : public ServerThread {
+  /*
+   * latch of the two maps above.
+   */
+  std::shared_mutex bLRPop_blocking_map_latch_;
```
blockMtx

cheniujh committed:
temp save

cheniujh committed:
improved code based on reviwer's opinion

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into new_add_blpop_brpop

cheniujh committed:
renamed some variables

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into new_add_blpop_brpop

cheniujh committed:
Merge remote-tracking branch 'origin/new_add_blpop_brpop' into new_add_blpop_brpop

cheniujh committed:
1. added record lock in ServeAndUnblockConns(void* args)
2. deleted testing code
3. formated code using clang-format

cheniujh committed:
1. added record lock in ServeAndUnblockConns(void* args)
2. deleted testing code
3. formated code using clang-format
4. removed kCmdFlagsMayDeferWriteBinlog and override dobinlog in blpop and brpop

wanghenshui commented:
> 主从复制测试（备注：rpoplpush命令本身主从同步有问题，所以测试里把对应部分先注释了） 

你先把这个修了吧

cheniujh committed:
added multi-db concurrency test

cheniujh committed:
renamed a flag from kCmdFlagsMayDfferWrite to kCmdFlagsMayDfferWriteBinlog

cheniujh committed:
Merge remote-tracking branch 'origin/new_add_blpop_brpop' into new_add_blpop_brpop

# Conflicts:
#	include/pika_command.h
#	src/pika_client_conn.cc
#	src/pika_command.cc
#	src/pika_list.cc

cheniujh committed:
removed unsed code

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +80,118 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
   }
 }
 
+struct BlrPopUnblockTaskArgs {
```
UnblockTaskArgs

wanghenshui replied:
这个参数可以单独提出来放到另外的头文件，后续bzpop也可以用到

cheniujh replied:
好的，塞到pika_command.h里了

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +80,118 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+  std::unique_lock latch(dispatchThread->GetBlockMtx());
+  auto it = key_to_conns_.find(blrPop_key);
+  if (it == key_to_conns_.end()) {
+    slot->LockMgr()->UnLock(key);
```
这个可以用defer 来封装，后面的unlock就可以省掉了

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +80,118 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+  }
+  CmdRes res;
+  std::vector<WriteBinlogOfPopArgs> pop_binlog_args;
+  auto& waitting_list_of_this_key = it->second;
```
名字太长了 waitting_list就行

wanghenshui commented on src/pika_list.cc:
```diff
@@ -76,6 +80,118 @@ void LLenCmd::Do(std::shared_ptr<Slot> slot) {
+void BlockingBaseCmd::WriteBinlogOfPop(std::vector<WriteBinlogOfPopArgs>& pop_args) {
+  // write binlog of l/rpop
+  for (auto& pop_arg : pop_args) {
+    std::shared_ptr<Cmd> pop_cmd;
```
这个没必要shared_ptr

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into new_add_blpop_brpop

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into new_add_blpop_brpop

cheniujh committed:
revised code based on opinion of reviewer

cheniujh committed:
Merge branch 'unstable_new' of github.com:cheniujh/pika into new_add_blpop_brpop

# Conflicts:
#	include/pika_list.h
#	src/pika_command.cc
#	src/pika_list.cc

wanghenshui reviewed (commented):
把测试代码集成到流水线里

cheniujh committed:
Merge branch 'unstable' of github.com:cheniujh/pika into new_add_blpop_brpop

# Conflicts:
#	include/pika_list.h

cheniujh committed:
add the unit test to github workflow

cheniujh committed:
revised unit test file

cheniujh committed:
removed time counting

AlexStocks commented on include/pika_command.h:
```diff
@@ -398,6 +401,18 @@ class CmdRes {
+struct UnblockTaskArgs {
+  std::string key;
+  std::shared_ptr<Slot> slot;
+  net::DispatchThread* dispatchThread;
```
net::DispatchThread* dispatchThread = nullptr;

AlexStocks commented on include/pika_list.h:
```diff
@@ -209,6 +254,26 @@ class LTrimCmd : public Cmd {
+  BRPopCmd(const std::string& name, int arity, uint16_t flag) : BlockingBaseCmd(name, arity, flag){};
+  virtual std::vector<std::string> current_key() const {
+    std::vector<std::string> res = keys_;
+    return res;
```
为啥不是直接 return keys_?

cheniujh replied:
确实应该改成return {keys}

cheniujh committed:
clear github action cache

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into new_add_blpop_brpop

cheniujh committed:
revised based on reviwer's opinions

cheniujh committed:
Merge branch 'OpenAtomFoundation:unstable' into new_add_blpop_brpop

cheniujh committed:
Modify to adapt to Mac.

cheniujh committed:
Merge remote-tracking branch 'origin/new_add_blpop_brpop' into new_add_blpop_brpop

cheniujh committed:
Modify to adapt to Mac 2.
